File: motio-commons/src/main/java/com/motio/commons/model/ShoppingList.java
package com.motio.commons.model;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "shopping_list")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
public class ShoppingList {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String listName;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    private List<ShoppingItem> items = new LinkedList<>();

    @ManyToOne
    @JoinColumn(name = "created_by_user_id", nullable = false)
    private User createdByUser;

    @ManyToMany
    @JoinTable(
            name = "shopping_list_access",
            joinColumns = @JoinColumn(name = "shopping_list_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private Set<User> accessibleUsers = new HashSet<>();
}

File: motio_client/lib/models/shopping_list.dart
import 'shopping_item.dart';
import 'user.dart';

class ShoppingList {
  final int id;
  final String listName;
  final List<ShoppingItem> items;
  final User createdByUser;
  final List<User> accessibleUsers;

  ShoppingList({
    required this.id,
    required this.listName,
    required this.items,
    required this.createdByUser,
    required this.accessibleUsers,
  });

  factory ShoppingList.fromJson(Map<String, dynamic> json) {
    return ShoppingList(
      id: json['id'],
      listName: json['listName'],
      items: (json['items'] as List).map((i) => ShoppingItem.fromJson(i)).toList(),
      createdByUser: User.fromJson(json['createdByUser']),
      accessibleUsers: (json['accessibleUsers'] as List).map((u) => User.fromJson(u)).toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'listName': listName,
      'items': items.map((i) => i.toJson()).toList(),
      'createdByUser': createdByUser.toJson(),
      'accessibleUsers': accessibleUsers.map((u) => u.toJson()).toList(),
    };
  }

  ShoppingList copyWith({
    int? id,
    String? listName,
    List<ShoppingItem>? items,
    User? createdByUser,
    List<User>? accessibleUsers,
  }) {
    return ShoppingList(
      id: id ?? this.id,
      listName: listName ?? this.listName,
      items: items ?? this.items,
      createdByUser: createdByUser ?? this.createdByUser,
      accessibleUsers: accessibleUsers ?? this.accessibleUsers,
    );
  }
}
